clang/include/clang/CIR/Dialect/IR/CIRTypes.td

```diff
@@ -603,11 +603,163 @@ def FuncPtr : Type<
+      "bool":$packed,
+      "bool":$padded,
+      "RecordKind":$kind,
+      CArg<"ASTRecordDeclInterface", "{}">:$ast
+    ), [{
+      return $_get($_ctxt, members, name, /*incomplete=*/false, packed, padded,
+                       kind, ast);    
+    }]>,
+
+    // Create an identified and incomplete struct type.
+    TypeBuilder<(ins
+      "mlir::StringAttr":$name,
+      "RecordKind":$kind
+    ), [{
+      return $_get($_ctxt, /*members=*/llvm::ArrayRef<Type>{}, name,
+                         /*incomplete=*/true, /*packed=*/false, /*padded=*/false,
+                         kind,
+                         /*ast=*/ASTRecordDeclInterface{});    
+    }]>,
+
+    // Create an anonymous struct type (always complete).
+    TypeBuilder<(ins
+      "llvm::ArrayRef<mlir::Type>":$members,
+      "bool":$packed,
+      "bool":$padded,
+      "RecordKind":$kind,
+      CArg<"ASTRecordDeclInterface", "{}">:$ast
+    ), [{
+      return $_get($_ctxt, members, mlir::StringAttr{}, /*incomplete=*/false, packed,
+                      padded, kind, ast);
+    }]>];
+
+  let extraClassDeclaration = [{
+    using Base::verifyInvariants;
+
+    enum RecordKind : uint32_t { Class, Union, Struct };
+
+    bool isClass() const { return getKind() == RecordKind::Class; };
+    bool isStruct() const { return getKind() == RecordKind::Struct; };
+    bool isUnion() const { return getKind() == RecordKind::Union; };
+    bool isComplete() const { return !isIncomplete(); };
+    bool isIncomplete() const;
+
+    void dropAst();
+
+    mlir::Type getLargestMember(const mlir::DataLayout &dataLayout) const;
+    size_t getNumElements() const { return getMembers().size(); };
+    std::string getKindAsStr() {
+      switch (getKind()) {
+      case RecordKind::Class:
+        return "class";
+      case RecordKind::Union:
+        return "union";
+      case RecordKind::Struct:
+        return "struct";
+      }
+      llvm_unreachable("Invalid value for StructType::getKind()");
+    }
+    std::string getPrefixedName() {
+      return getKindAsStr() + "." + getName().getValue().str();
+    }
+
+    void complete(llvm::ArrayRef<mlir::Type> members, bool packed, bool isPadded,
+              ASTRecordDeclInterface ast = {});
+
+    uint64_t getElementOffset(const mlir::DataLayout &dataLayout,
+              unsigned idx) const;
+
+    bool isLayoutIdentical(const StructType &other);
+
+  // Utilities for lazily computing and cacheing data layout info.
+  // FIXME: currently opaque because there's a cycle if CIRTypes.types include
+  // from CIRAttrs.h. The implementation operates in terms of StructLayoutAttr
+  // instead.
+  private:
+    mutable mlir::Attribute layoutInfo;
+    void computeSizeAndAlignment(const mlir::DataLayout &dataLayout) const;
+  public:
+  }];
+
+  let hasCustomAssemblyFormat = 1;
+}
+
+// This type is used because of some limitations of type constraints.
+// See https://github.com/llvm/clangir/pull/1504#issuecomment-2738968751 for more details.
```

Comment:
Reword: "Note CIRStructType is used instead of CIR_StructType because of tablegen conflicts", no need for links